src/safeposix/syscalls/fs_calls.rs

```diff
@@ -2388,4 +2438,148 @@ impl Cage {
             syscall_error(Errno::EBADF, "cond_wait", "Condvar handle does not refer to a valid condvar!")
         }
     }
+
+    //------------------SEMAPHORE SYSCALLS------------------
+    /* 
+    *  Initialize semaphore object SEM to value
+    *  pshared used to indicate whether the semaphore is shared in threads (when equals to 0)
+    *  or shared between processes (when nonzero)
+    */
+    pub fn sem_init_syscall(&self, sem_handle: u32, pshared: i32, value: u32) -> i32 {
```

Comment:
What is `sem_handle`? Is it  just like a struct that holds the shared memory addresses and as well as whatever threads or cages are referencing that semaphore?

Reply:
im not sure if I understand your question well, I think `sem_handle` is the semaphore ptr. NaCl will send the semaphore ptr and cast that to u32. We add a dashmap called `sem_table` in Cage struct, mapping address of semaphore to semaphore struct we created in safeposix. and in RustSemaphore we hold the value of sem and whether it could be shared between processes.

Reply:
Ah okay, that makes sense, thanks.